server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java

```diff
@@ -312,16 +316,23 @@ public void messageReceived(RecoveryFinalizeRecoveryRequest request, TransportCh
         public void messageReceived(final RecoveryHandoffPrimaryContextRequest request, final TransportChannel channel,
                                     Task task) throws Exception {
             final RecoveryRef recoveryRef = onGoingRecoveries.getRecoverySafe(request.recoveryId(), request.shardId());
+            final List<Releasable> toRelease = new ArrayList<>(2);
+            toRelease.add(recoveryRef::close);
             boolean success = false;
             try {
+                // Due to relocation conditions on the shard it could take a while for the hand-off to complete so we disable the recovery
+                // monitor since we don't expect any transport messages from master for the duration of the handoff and activate it again
+                // after the handoff.
+                final Releasable disabledMonitor = recoveryRef.target().disableRecoveryMonitor();
```

Comment:
This is a bit of a BwC issue I guess. If the hand-off request comes from `7.10` and doesn't wait indefinitely yet then it could timeout on the primary I guess.